src/poprox_recommender/components/diversifiers/mmr.py

```diff
@@ -46,24 +42,26 @@ def mmr_diversification(rewards, similarity_matrix, theta: float, topk: int):
     S.append(rewards.argmax())
 
     for k in range(topk - 1):
-        candidate = None  # next item
-        best_MR = float("-inf")
-
-        for i, reward_i in enumerate(rewards):  # iterate R for next item
-            if i in S:
-                continue
-            max_sim = float("-inf")
-
-            for j in S:
-                sim = similarity_matrix[i][j]
-                if sim > max_sim:
-                    max_sim = sim
-
-            mr_i = theta * reward_i - (1 - theta) * max_sim
-            if mr_i > best_MR:
-                best_MR = mr_i
-                candidate = i
-
-        if candidate is not None:
-            S.append(candidate)
+        # find the best combo of reward and max sim to existing item
+        # first, let's pare the matrix: candidates on rows, selected items on cols
+        M = similarity_matrix[:, S]
+
+        # for each target item, we want to find the *max* simialrity to an existing.
+        # we do this by taking the min of each row.
+        scores = torch.max(M, axis=1)
+        assert len(scores) == len(rewards)
+
+        # now, we want to compute θ*r - (1-θ)*s. let's do that *in-place* using
+        # this scores vector. To start, multiply by θ-1 (-(1-θ)):
+        scores *= theta - 1
+
+        # with this, we can theta * rewards in-place:
+        scores.add_(rewards, theta)
```

Comment:
Possible to name the arguments here? Not obvious how this is `theta * rewards`

Reply:
Not sure how much it will help readability, since their names are `other` and `alpha`, but we actually do need to name the second one — it's keyword-only.

Reply:
the comment should have another word, though — "we can add theta * rewards"